cleverhans/attacks_tf.py

```diff
@@ -417,3 +417,245 @@ def jacobian_augmentation(sess, x, X_sub_prev, Y_sub, grads, lmbda,
 
     # Return augmented training data (needs to be labeled afterwards)
     return X_sub
+
+
+class CarliniWagnerL2:
```

Comment:
Okay, so I think that actually having the second class is useful. Since the attack is implemented in python, and generate() is implemented with a py_func, if the py_func created a second instance of the class, then the pseudo-code

    args = {...}
    xs = CarliniWagnerL2().generate(...)
    a = sess.run(xs, args)
    b = sess.run(xs, args)

would result in constructing the graph twice. JSMA has this problem too, but gets around it by placing a call to

    grads = jacobian_graph(preds, x, self.nb_classes)  

outside of the py_func. This is equivelant to what I do now, which is instantiate the second class outside of the py_func. In my case, it's easier to have a second class instead of a helper function because there are so many elements of the graph that are re-used and just saved as instance variables.